Cgen does not generate `map[T].clear()` methods in generic structs

### Describe the bug

<title>

### Reproduction Steps

```v
pub type EventListener[T] = fn (T) !

pub struct EventController[T] {
mut:
	id        int
	listeners map[int]EventListener[T]
}

fn (mut ec EventController[T]) generate_id() int {
	return ec.id++
}

pub fn (mut ec EventController[T]) override(listener EventListener[T]) EventController[T] {
	ec.listeners.clear()
	return ec.listen(listener)
}

pub fn (mut ec EventController[T]) listen(listener EventListener[T]) EventController[T] {
	ec.listeners[ec.generate_id()] = listener
	return ec
}

struct Foo {}
struct Bar {}

fn make[T](i int) EventController[T]{
	return EventController[T]{id: i}
}

fn main() {
	assert make[Foo](2) == EventController[Foo]{id: 2}
	assert make[Bar](5) == EventController[Bar]{id: 5}
}
```

### Expected Behavior

Successful compile, and execution

### Current Behavior

```
==================
C:/Users/mclr/AppData/Local/Temp/v_0/test.15956463521132725052.tmp.c:12996: warning: implicit declaration of function 'Map_int_anon_fn_main__bar__result_void_clear'
tcc: error: undefined symbol 'Map_int_anon_fn_main__foo__result_void_clear'
tcc: error: undefined symbol 'Map_int_anon_fn_main__bar__result_void_clear'
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose
```

### Possible Solution

_No response_

### Additional Information/Context

just making [`discord.EventController#override`](https://github.com/DarpHome/discord.v/blob/659d44ea2815df9c7f92aee1f83ae052418c6f6c/discord/eventcontroller.v#L149-L153)

### V version

V 0.4.3 ae82b67

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.3 57a33c3.ae82b67
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 12 cpus, 64bit, little endian, 

getwd: D:\Games\Proekti\V\interactions
vexe: D:\Games\Proekti\V\v\v.exe
vexe mtime: 2024-01-01 21:47:17

vroot: OK, value: D:\Games\Proekti\V\v
VMODULES: OK, value: C:\Users\mclr\.vmodules
VTMP: OK, value: C:\Users\mclr\AppData\Local\Temp\v_0

Git version: git version 2.37.0.windows.1
Git vroot status: weekly.2023.52-45-gae82b67e
.git/config present: true

CC version:
thirdparty/tcc status: thirdparty-windows-amd64 a39eb79b
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -880,8 +880,6 @@ fn (mut g Gen) gen_arg_from_type(node_type ast.Type, node ast.Expr) {
 fn (mut g Gen) gen_map_method_call(node ast.CallExpr, left_type ast.Type, left_sym ast.TypeSymbol) bool {
 	match node.name {
 		'clear' {
-			left_info := left_sym.info as ast.Map
-			elem_type_str := g.typ(left_info.key_type)
 			g.write('map_clear(')
 			g.gen_arg_from_type(left_type, node.left)
 			g.write(')')

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -879,6 +879,14 @@ fn (mut g Gen) gen_arg_from_type(node_type ast.Type, node ast.Expr) {
 
 fn (mut g Gen) gen_map_method_call(node ast.CallExpr, left_type ast.Type, left_sym ast.TypeSymbol) bool {
 	match node.name {
+		'clear' {
+			left_info := left_sym.info as ast.Map
+			elem_type_str := g.typ(left_info.key_type)
+			g.write('map_clear(')
+			g.gen_arg_from_type(left_type, node.left)
+			g.write(')')
+			return true
+		}
 		'delete' {
 			left_info := left_sym.info as ast.Map
 			elem_type_str := g.typ(left_info.key_type)

--- a/vlib/v/tests/map_clear_test.v
+++ b/vlib/v/tests/map_clear_test.v
@@ -1,0 +1,49 @@
+pub type EventListener[T] = fn (T) !
+
+pub struct EventController[T] {
+mut:
+	id        int
+	listeners map[int]EventListener[T]
+}
+
+fn (mut ec EventController[T]) generate_id() int {
+	return ec.id++
+}
+
+pub fn (mut ec EventController[T]) override(listener EventListener[T]) EventController[T] {
+	ec.listeners.clear()
+	return ec.listen(listener)
+}
+
+pub fn (mut ec EventController[T]) listen(listener EventListener[T]) EventController[T] {
+	ec.listeners[ec.generate_id()] = listener
+	return ec
+}
+
+struct Foo {}
+
+struct Bar {}
+
+fn make[T](i int) EventController[T] {
+	return EventController[T]{
+		id: i
+	}
+}
+
+fn test_main() {
+	mut a := EventController[Foo]{
+		id: 1
+		listeners: {
+			1: fn (a Foo) ! {
+				dump(1)
+			}
+		}
+	}
+	assert dump(a).id == 1
+	assert dump(a).listeners.len == 1
+	a.override(fn (a Foo) ! {
+		dump(2)
+	})
+	assert dump(a).id == 2
+	assert dump(a).listeners.len == 1
+}

